pkg/controller/daemon/util/daemonset_util.go

```diff
@@ -133,3 +134,60 @@ func SplitByAvailablePods(minReadySeconds int32, pods []*v1.Pod) ([]*v1.Pod, []*
 	}
 	return availablePods, unavailablePods
 }
+
+// AppendDaemonSetPodNodeAffinity adds a NodeAffinity term with the given "nodeName" to the "affinity" terms.
+func AppendDaemonSetPodNodeAffinity(affinity *v1.Affinity, nodename string) *v1.Affinity {
+	nodeSelector := &v1.NodeSelector{
+		NodeSelectorTerms: []v1.NodeSelectorTerm{
+			{
+				MatchExpressions: []v1.NodeSelectorRequirement{
+					{
+						Key:      kubeletapis.LabelHostname,
+						Operator: v1.NodeSelectorOpIn,
+						Values:   []string{nodename},
+					},
+				},
+			},
+		},
+	}
+
+	if affinity == nil {
+		return &v1.Affinity{
+			NodeAffinity: &v1.NodeAffinity{
+				RequiredDuringSchedulingIgnoredDuringExecution: nodeSelector,
+			},
+		}
+	}
+
+	if affinity.NodeAffinity == nil {
+		affinity.NodeAffinity = &v1.NodeAffinity{
+			RequiredDuringSchedulingIgnoredDuringExecution: nodeSelector,
+		}
+		return affinity
+	}
+
+	nodeAffinity := affinity.NodeAffinity
+
+	if nodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution == nil {
+		nodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution = nodeSelector
+		return affinity
+	}
+
+	// Removes hostname node selector, as only the target hostname will take effect.
+	for _, term := range nodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms {
+		exps := []v1.NodeSelectorRequirement{}
+		for _, exp := range term.MatchExpressions {
+			if exp.Key != kubeletapis.LabelHostname {
+				exps = append(exps, exp)
+			}
+		}
+		term.MatchExpressions = exps
```

Comment:
How does the default scheduler handle affinity conflicts? For example, when the node affinity is specified like this:

1. Schedule on node with label `hostname in [foo]`, and 
1. Schedule on node with label `hostname in [bar]`

Or this case:

1. Schedule on node with label `hostname in [foo]`, and
1. Schedule on node with label `hostname not in [foo]`

Reply:
In predicates, we use NodeAffinity to identify the nodes we're going to run DaemonSet. So the hostname in this function is one of NodeAffinity.

For your cases, the `nodeSelectionPredicates` will return empty, so it'll not happen :)